deepchem/metrics/__init__.py

```diff
@@ -270,38 +543,60 @@ def __init__(self,
                name=None,
                threshold=None,
                mode=None,
-               compute_energy_metric=False):
+               compute_energy_metric=None):
     """
     Parameters
     ----------
     metric: function
-      function that takes args y_true, y_pred (in that order) and
-      computes desired score.
-    task_averager: function, optional
+      Function that takes args y_true, y_pred (in that order) and
+      computes desired score. If sample weights are to be considered,
+      `metric` may take in an additional keyword argument
+      `sample_weight`.
+    task_averager: function, optional (default, np.mean)
       If not None, should be a function that averages metrics across
-      tasks. For example, task_averager=np.mean. If task_averager is
-      provided, this task will be inherited as a multitask metric.
-    name: str, optional
+      tasks. 
+    name: str, optional (default None)
       Name of this metric
-    threshold: float, optional
+    threshold: float, optional (default None) (DEPRECATED)
       Used for binary metrics and is the threshold for the positive
-      class
-    mode: str, optional
-      Must be either classification or regression.
-    compute_energy_metric: TODO(rbharath): Should this be removed? 
+      class.
+    mode: str, optional (default None)
+      Should usually be "classification" or "regression."
+    compute_energy_metric: bool, optional (default None) (DEPRECATED)
+      Deprecated metric. Will be removed in a future version of
+      DeepChem. Do not use.
     """
+    if threshold is not None:
+      logger.warn(
+          "threshold is deprecated and will be removed in a future version of DeepChem. Set threshold in compute_metric instead"
+      )
+    if compute_energy_metric is not None:
+      self.compute_energy_metric = compute_energy_metric
+      logger.warn(
+          "compute_energy_metric is deprecated and will be removed in a future version of DeepChem."
+      )
+    else:
+      self.compute_energy_metric = False
     self.metric = metric
-    self.task_averager = task_averager
-    self.is_multitask = (self.task_averager is not None)
+    if task_averager is None:
+      self.task_averager = np.mean
+    else:
+      self.task_averager = task_averager
     if name is None:
-      if not self.is_multitask:
-        self.name = self.metric.__name__
+      if task_averager is None:
+        if hasattr(self.metric, '__name__'):
+          self.name = self.metric.__name__
+        else:
+          self.name = "unknown metric"
       else:
-        self.name = self.task_averager.__name__ + "-" + self.metric.__name__
+        if hasattr(self.metric, '__name__'):
+          self.name = task_averager.__name__ + "-" + self.metric.__name__
+        else:
+          self.name = "unknown metric"
     else:
       self.name = name
```

Comment:
Good catch! Will clean this up

Reply:
Looking at this more closely, I think the original structure is needed. The main reason is we're casing on the argument `task_averager` (which may still be `None`) rather than `self.task_averager` which is guaranteed to not be `None`. This handling is needed to handle singletask metrics correctly.